client/src/main/java/org/evosuite/ga/archive/MIOArchive.java

```diff
@@ -0,0 +1,575 @@
+          this.solutions.add(candidateSolution);
+          this.sortPairSolutions(); // keep solutions sorted from the best to the worse
+        } else {
+          // no, there is not. so, replace the worst one, if candidate is better.
+          this.sortPairSolutions();
+          Pair<Double, T> worstSolution = this.solutions.get(this.capacity - 1);
+
+          if (isPairBetterThanCurrent(worstSolution, candidateSolution)) {
+            this.solutions.set(this.capacity - 1, candidateSolution);
+          }
+        }
+      }
+
+      // a set of solutions larger that a maximum capacity would be considered illegal
+      assert this.solutions.size() <= this.capacity;
+
+      if (added) {
+        // reset counter if and only if a new/better solution has been found
+        this.counter = 0;
+      }
+
+      return added;
+    }
+
+    /**
+     * 
+     * @param currentSolution
+     * @param candidateSolution
+     * @return
+     */
+    private boolean isPairBetterThanCurrent(Pair<Double, T> currentSolution,
+        Pair<Double, T> candidateSolution) {
+      int cmp = Double.compare(currentSolution.getLeft(), candidateSolution.getLeft());
+      if (cmp < 0) {
+        return true;
+      } else if (cmp > 0) {
+        return false;
+      }
+      assert cmp == 0;
+
+      return isBetterThanCurrent(currentSolution.getRight(), candidateSolution.getRight());
+    }
+
+    /**
+     * 
+     * @return
+     */
+    private T sampleSolution() {
+      if (this.numSolutions() == 0) {
+        return null;
+      }
+      this.counter++;
+      return Randomness.choice(this.solutions).getRight();
+    }
+
+    /**
+     * DESC sort, i.e., from the pair with the highest h to the pair with the lowest h
+     */
+    private void sortPairSolutions() {
+      this.solutions.sort(new Comparator<Pair<Double, T>>() {
+        @Override
+        public int compare(Pair<Double, T> solution0, Pair<Double, T> solution1) {
+          if (solution0.getLeft() < solution1.getLeft()) {
+            return 1;
+          } else if (solution0.getLeft() > solution1.getLeft()) {
+            return -1;
+          }
+          return 0;
+        }
+      });
+    }
+
+    /**
+     * 
+     * @return
+     */
+    private int numSolutions() {
+      return this.solutions.size();
+    }
+
+    /**
+     * 
+     * @return
+     */
+    private T getBestSolutionIfAny() {
+      if (this.numSolutions() == 0 || !this.isCovered()) {
+        return null;
+      }
+      return this.solutions.get(0).getRight();
+    }
+
+    /**
+     * 
+     * @param newPopulationSize
+     */
+    private void shrinkPopulation(int newPopulationSize) {
+      assert newPopulationSize > 0;
+
+      if (this.isCovered()) {
+        return;
+      }
+
+      this.capacity = newPopulationSize;
+
+      if (this.numSolutions() < newPopulationSize) {
+        // no need to shrink it
+        return;
+      }
+
```

Comment:
when u are shrinking here, are you making sure that the deleted solutions are the worst in that population?

Reply:
Yes. Population is sorted (best to worst) every time we update it (see lines 421 and 424), so that we can easily shrink or replace the worst one(s) at any time. Although the sort is performed frequently, I don't expect it to be expensive as the size of a population is typically very small.